src/odemis/acq/stream/_projection.py

```diff
@@ -218,6 +224,428 @@ def _updateImage(self):
+
+                # convert data into rectangular format (theta-phi-representation)
+                # Check if rectangular converted representation already was calculated for requested ebeam pos.
+                # Note: This calc is very time consuming. Takes about 3.6 sec for one ebeam pos (conversion of 6 images)
+                # and tested on an image of size (256, 1024)
+
+                if ebeam_pos not in polarimetry_cache_raw.keys():
+                    # Get the center wavelength of the filter used (no filter aka "pass-through" use fallback)
+                    # Does not matter from which of the 6 images as they all must have been recorded with the same
+                    # filter -> reuse key
+                    # TODO don't like this solution, key will be the last value of key in the for loop above
+                    # the value of key might be a different one each time...and it is only to get the wl of the filter.
+                    # not sure this is still working in python 3 as I learned today ;)
+                    fallback_wl = 650e-9
+                    band = data_raw_bg_corr[key].metadata.get(model.MD_OUT_WL, fallback_wl)
+                    if isinstance(band, tuple):  # wl is usually tuple of min/max value
+                        wl = sum(band) / len(band)
+                    else:  # handles if band is str
+                        # TODO if type is "str", support center wavelength based on color
+                        wl = fallback_wl
+
+                    # Warning: allocates lot of memory, which will not be free'd until
+                    # the current thread is terminated.
+
+                    # TODO allow variable input size? Calc based on raw data?
+                    size = (180, 720)  # (2*90, 2*360) defines the resolution of the displayed image
+
+                    for key in data_raw_bg_corr.keys():
+                        data_raw_bg_corr[key] = angleres.AngleResolved2Rectangular(data_raw_bg_corr[key], size)
+                    polarimetry_cache_raw[ebeam_pos] = data_raw_bg_corr  # dict containing the bg corrected raw data
+
+                    # Calculate the polarimetry results:
+                    # Dictionary keys are (ebeam pos x, ebeam pos y, optional: polarimetry key).
+                    # Note: Takes about 0.25 sec to calc all polarimetry results for one ebeam pos
+                    # and tested on an image of size (256, 1024)
+                    polarimetry_cache_raw.update(
+                        arpolvis.calcPolarimetry(polarimetry_cache_raw, ebeam_pos, size, wl))
+
+                # Convert the data to polar representation for GUI display.
+                # Note: Takes 0.24 sec to convert one image for display and tested on an image of size (256, 1024)
+                # TODO invalid division check if only bg  # norm so far
+                # TODO move to update image?? as this is the only code we need to reuse
```

Comment:
Indeed, I think it'd be clearer, and be more reusable if this function is broken into two parts: 1) compute the stoke raw data, and 2) project to RGB. In particular, for each step there is a corresponding cache, so that would clear up the differences between the two caches.
It might also help when you add `_projectAsRaw()`.